File: rebel/solver.py
"""Perfect-information ("double-dummy") solver for the Euchre play phase.

Given a fully-specified state (all four hands known), compute the value under
optimal play by both teams, where value is ``team0_points - team1_points`` for
the hand. This is exact game-tree search with alpha-beta pruning, a
transposition table, and move ordering, which makes five-card double-dummy
solves fast enough to sit inside PIMC's per-decision world loop.

It is a *team* game rather than a strictly alternating one (partners share an
objective), but it is still one scalar that team 0 maximizes and team 1
minimizes, so alpha-beta applies.

The solver is the engine behind PIMC search (`rebel/pimc.py`): sample many
worlds consistent with what a player knows, solve each, and aggregate.
"""

from __future__ import annotations

from collections import defaultdict
from typing import Dict, List, Optional, Tuple

from euchre.actions import Play, Action
from euchre.cards import Card, card_strength, effective_suit
from euchre.game import EuchreState, Phase, team_of

_INF = 10 ** 9
_LOWER, _EXACT, _UPPER = 0, 1, 2  # transposition-table bound flags


def _key(state: EuchreState) -> Tuple:
    hands = tuple(tuple(sorted(c.id for c in state.hands[p])) for p in range(4))
    trick = tuple((seat, c.id) for seat, c in state.current_trick)
    return (state.current_player, trick, hands,
            state.tricks_won[0], state.tricks_won[1])


def _ordered_plays(state: EuchreState, player: int) -> List[Card]:
    """Legal plays, de-duplicated by equivalence and ordered strongest-first.

    Two of a player's cards in the same effective suit are strategically
    equivalent when they form a consecutive run among the still-unplayed cards
    (no opponent card ranks between them): playing either yields the same
    double-dummy value, so we keep a single representative. This is the classic
    double-dummy move reduction and it slashes the branching of the first
    tricks. The remaining moves are then ordered strongest-first for pruning.
    """
    plays = state._legal_plays(player)
    if len(plays) <= 1:
        return plays
    trump = state.trump
    myset = set(state.hands[player])
    playset = set(plays)

    # Bucket cards by effective suit, ordered strongest-first. Cards already
    # played to the *current* trick must be included: they are not playable but
    # they still separate ranks (a card sitting in the trick can rank between
    # two of my cards, so playing the higher wins the trick while the lower
    # does not -- those two are NOT equivalent).
    buckets: Dict[object, List[Card]] = defaultdict(list)
    for p in range(4):
        for c in state.hands[p]:
            buckets[effective_suit(c, trump)].append(c)
    for _seat, c in state.current_trick:
        buckets[effective_suit(c, trump)].append(c)

    keep: List[Card] = []
    for suit, cards in buckets.items():
        cards.sort(key=lambda c: card_strength(c, trump, suit), reverse=True)
        i, n = 0, len(cards)
        while i < n:
            if cards[i] in myset:
                run = []
                while i < n and cards[i] in myset:
                    run.append(cards[i])
                    i += 1
                legal_in_run = [c for c in run if c in playset]
                if legal_in_run:
                    keep.append(legal_in_run[-1])  # lowest legal representative
            else:
                i += 1

    reduced = keep if keep else plays
    if state.current_trick:
        led = effective_suit(state.current_trick[0][1], trump)
    else:
        led = None
    return sorted(
        reduced,
        key=lambda c: card_strength(c, trump, led if led is not None else c.suit),
        reverse=True,
    )


def _ab(state: EuchreState, alpha: int, beta: int,
        memo: Dict[Tuple, Tuple[int, int]]) -> int:
    if state.is_terminal():
        r = state.returns()
        return r[0] - r[1]

    a0, b0 = alpha, beta
    key = _key(state)
    entry = memo.get(key)
    if entry is not None:
        val, flag = entry
        if flag == _EXACT:
            return val
        if flag == _LOWER:
            alpha = max(alpha, val)
        else:  # _UPPER
            beta = min(beta, val)
        if alpha >= beta:
            return val

    player = state.current_player
    maximizing = team_of(player) == 0
    plays = _ordered_plays(state, player)

    if maximizing:
        value = -_INF
        for c in plays:
            value = max(value, _ab(state.apply(Play(c)), alpha, beta, memo))
            alpha = max(alpha, value)
            if alpha >= beta:
                break
    else:
        value = _INF
        for c in plays:
            value = min(value, _ab(state.apply(Play(c)), alpha, beta, memo))
            beta = min(beta, value)
            if alpha >= beta:
                break

    if value <= a0:
        flag = _UPPER
    elif value >= b0:
        flag = _LOWER
    else:
        flag = _EXACT
    memo[key] = (value, flag)
    return value


def solve_value(state: EuchreState,
                memo: Optional[Dict[Tuple, Tuple[int, int]]] = None) -> int:
    """Optimal ``team0 - team1`` point differential from ``state`` (PLAY phase).

    ``memo`` may be shared across sibling calls to reuse transpositions.
    """
    if state.is_terminal():
        r = state.returns()
        return r[0] - r[1]
    assert state.phase == Phase.PLAY, "solver runs on the play phase"
    if memo is None:
        memo = {}
    return _ab(state, -_INF, _INF, memo)


def action_values(state: EuchreState,
                  memo: Optional[Dict[Tuple, Tuple[int, int]]] = None
                  ) -> Dict[Action, int]:
    """Optimal continuation value (team0 - team1) for each legal play.

    Each child is solved with a full window so the returned values are exact;
    a shared transposition table keeps sibling solves cheap.
    """
    assert state.phase == Phase.PLAY
    if memo is None:
        memo = {}
    out: Dict[Action, int] = {}
    for c in _ordered_plays(state, state.current_player):
        out[Play(c)] = _ab(state.apply(Play(c)), -_INF, _INF, memo)
    return out


def best_play(state: EuchreState,
              memo: Optional[Dict[Tuple, Tuple[int, int]]] = None) -> Action:
    """The optimal play for the side on move (from its own team's view)."""
    maximizing = team_of(state.current_player) == 0
    values = action_values(state, memo)
    return (max if maximizing else min)(values, key=lambda a: values[a])

File: cpp/solver.cpp
#include "solver.h"

#include <algorithm>
#include <sstream>

namespace mceuchre {

namespace {

constexpr int INF = 1'000'000'000;
enum Bound { LOWER = 0, EXACT = 1, UPPER = 2 };

// Transposition-table key. Only needs to be internally consistent (never
// compared cross-language), so plain string-building is fine -- correctness
// of solve_value's returned VALUE doesn't depend on this format at all.
std::string tt_key(const EuchreState& s) {
    std::ostringstream os;
    os << static_cast<int>(s.current_player) << "|";
    for (const auto& p : s.current_trick) os << static_cast<int>(p.player) << ":" << static_cast<int>(p.card) << ";";
    os << "|";
    for (int p = 0; p < 4; ++p) {
        auto ids = hand_to_vector(s.hands[p]);
        std::sort(ids.begin(), ids.end());
        for (CardId c : ids) os << static_cast<int>(c) << ",";
        os << "|";
    }
    os << s.tricks_won[0] << "," << s.tricks_won[1];
    return os.str();
}

// Move reduction + strongest-first ordering (mirrors solver.py's
// _ordered_plays). The exact tie-break order among strategically-equivalent
// alternatives doesn't affect solve_value's returned value (alpha-beta's
// result is move-order-invariant given a correct move SET, only pruning
// efficiency varies) -- only the kept representative SET matters for
// correctness, which is what this must get exactly right.
std::vector<CardId> ordered_plays(const EuchreState& state, int player) {
    auto plays = state.legal_plays(player);
    if (plays.size() <= 1) return plays;
    int trump = *state.trump;
    bool my[NUM_CARDS] = {false};
    for (CardId c : hand_to_vector(state.hands[player])) my[c] = true;
    bool legal[NUM_CARDS] = {false};
    for (CardId c : plays) legal[c] = true;

    // Bucket by effective suit (0..3): all 4 hands' cards + current-trick cards.
    std::array<std::vector<CardId>, NUM_SUITS> buckets;
    for (int p = 0; p < 4; ++p)
        for (CardId c : hand_to_vector(state.hands[p])) buckets[effective_suit(c, trump)].push_back(c);
    for (const auto& tp : state.current_trick) buckets[effective_suit(tp.card, trump)].push_back(tp.card);

    std::vector<CardId> keep;
    for (int suit = 0; suit < NUM_SUITS; ++suit) {
        auto& cards = buckets[suit];
        if (cards.empty()) continue;
        std::sort(cards.begin(), cards.end(), [&](CardId a, CardId b) {
            return card_strength(a, trump, suit) > card_strength(b, trump, suit);
        });
        size_t i = 0, n = cards.size();
        while (i < n) {
            if (my[cards[i]]) {
                std::vector<CardId> run;
                while (i < n && my[cards[i]]) { run.push_back(cards[i]); ++i; }
                CardId lowest_legal = -1;
                for (CardId c : run) if (legal[c]) lowest_legal = c;  // last legal in run
                if (lowest_legal >= 0) keep.push_back(lowest_legal);
            } else {
                ++i;
            }
        }
    }

    const std::vector<CardId>& reduced = keep.empty() ? plays : keep;
    int led = -1;
    bool has_led = !state.current_trick.empty();
    if (has_led) led = effective_suit(state.current_trick[0].card, trump);

    std::vector<CardId> out(reduced);
    std::sort(out.begin(), out.end(), [&](CardId a, CardId b) {
        int la = has_led ? led : card_suit(a);
        int lb = has_led ? led : card_suit(b);
        return card_strength(a, trump, la) > card_strength(b, trump, lb);
    });
    return out;
}

int ab(const EuchreState& state, int alpha, int beta,
      std::unordered_map<std::string, std::pair<int, int>>& memo) {
    if (state.is_terminal()) {
        auto [p0, p1] = state.returns();
        return p0 - p1;
    }

    int a0 = alpha, b0 = beta;
    std::string key = tt_key(state);
    auto it = memo.find(key);
    if (it != memo.end()) {
        auto [val, flag] = it->second;
        if (flag == EXACT) return val;
        if (flag == LOWER) alpha = std::max(alpha, val);
        else beta = std::min(beta, val);
        if (alpha >= beta) return val;
    }

    int player = state.current_player;
    bool maximizing = team_of(player) == 0;
    auto plays = ordered_plays(state, player);

    int value;
    if (maximizing) {
        value = -INF;
        for (CardId c : plays) {
            value = std::max(value, ab(state.apply(Action::play(c)), alpha, beta, memo));
            alpha = std::max(alpha, value);
            if (alpha >= beta) break;
        }
    } else {
        value = INF;
        for (CardId c : plays) {
            value = std::min(value, ab(state.apply(Action::play(c)), alpha, beta, memo));
            beta = std::min(beta, value);
            if (alpha >= beta) break;
        }
    }

    int flag = (value <= a0) ? UPPER : (value >= b0) ? LOWER : EXACT;
    memo[key] = {value, flag};
    return value;
}

}  // namespace

int solve_value(const EuchreState& state, std::unordered_map<std::string, std::pair<int, int>>* memo) {
    if (state.is_terminal()) {
        auto [p0, p1] = state.returns();
        return p0 - p1;
    }
    std::unordered_map<std::string, std::pair<int, int>> local;
    auto& m = memo ? *memo : local;
    return ab(state, -INF, INF, m);
}

}  // namespace mceuchre
